crates/core/tedge_api/src/entity_store.rs

```diff
@@ -939,114 +940,64 @@ pub enum InitError {
-
-        let payload = parse_entity_register_payload(message.payload_bytes())?;
-
-        let JsonValue::Object(mut properties) = payload else {
-            return None;
-        };
-
-        let Some(JsonValue::String(r#type)) = properties.remove("@type") else {
-            return None;
-        };
-
-        let r#type = match r#type.as_str() {
-            "device" => EntityType::MainDevice,
-            "child-device" => EntityType::ChildDevice,
-            "service" => EntityType::Service,
-            _ => return None,
-        };
-
-        let parent = properties.remove("@parent");
-        let parent = if let Some(parent) = parent {
-            let JsonValue::String(parent) = parent else {
-                return None;
-            };
-            let Ok(parent) = parent.parse() else {
-                return None;
-            };
-            Some(parent)
-        } else {
-            None
-        };
-
-        let parent = if r#type == EntityType::ChildDevice || r#type == EntityType::Service {
-            parent
-        } else {
-            None
-        };
-
-        let health_service = if let Some(value) = properties.remove("@health") {
-            let JsonValue::String(value_str) = value else {
-                return None;
-            };
-            let Ok(health_service) = value_str.parse() else {
-                return None;
-            };
-            Some(health_service)
-        } else {
-            None
-        };
-
-        let entity_id = properties.remove("@id");
-        let entity_id = if let Some(entity_id) = entity_id {
-            let JsonValue::String(entity_id) = entity_id else {
-                return None;
-            };
-            Some(entity_id.into())
-        } else {
-            None
-        };
-
-        let other = properties;
-
-        assert_eq!(other.get("@id"), None);
-        assert_eq!(other.get("@type"), None);
-        assert_eq!(other.get("@parent"), None);
-
-        Some(Self {
-            topic_id: topic_id.parse().ok()?,
-            external_id: entity_id,
-            r#type,
-            parent,
-            health_endpoint: health_service,
-            twin_data: other,
-        })
+            .and_then(|s| s.strip_prefix('/'))
+            .ok_or_else(|| {
+                serde::de::Error::custom("Failed to parse MQTT topic as a valid entity topic")
+            })?;
+        let topic_id = EntityTopicId::from_str(topic_id).map_err(|_| {
+            serde::de::Error::custom("Failed to parse MQTT topic as a valid entity topic")
+        })?;
+
+        let mut payload: EntityRegistrationPayload =
+            serde_json::from_slice(message.payload_bytes())?;
+        if payload.topic_id.is_some() {
+            return Err(serde::de::Error::custom(
+                "@topic-id should not be present in the MQTT payload as it must be specified in the topic",
+            ));
+        }
+        payload.topic_id = Some(topic_id.clone());
```

Comment:
Wouldn't it be simpler to remove the topic-id field from `EntityRegistrationPayload` as it must indeed never be used?

Sure, we can no more simply return `payload.try_into()` to get the expected `EntityRegistrationMessage`, but this would really better convey the intent.

Reply:
> Wouldn't it be simpler to remove the topic-id field from EntityRegistrationPayload as it must indeed never be used?

The `@topic-id` was added to this struct since the registration payload received via HTTP would have the topic-id in the payload and not the URL path. The other alternative would have been to drop the `@topic-id` from this struct and then introduce another struct specifically for HTTP which includes the `topic-id` and this `EntityRegistrationPayload` as a "flattened" field.

Reply:
Resolved by [8654090](https://github.com/thin-edge/thin-edge.io/pull/3587/commits/8654090f76cf664e56dff2f493927e9a1b324c18)